### PR TITLE
92 implement exec subshell

### LINKED ISSUE #92: Implement `exec_subshell`

### PR DESCRIPTION
**Pull request by rask24**
fix #92

- **Implement execute_subshell**
- **Fix io and wait**
- **Add e2e test for subshell**

## Summary by CodeRabbit

- **New Features**
	- Introduced subshell execution capabilities, enhancing command execution within the application.
	- Added new functionalities for handling I/O operations and command spawning.
	- Enhanced pipeline execution logic with improved error handling and resource management.

- **Bug Fixes**
	- Improved control flow to accommodate subshell execution, preventing potential errors related to null parameters.

- **Tests**
	- Added comprehensive tests for various shell command redirection scenarios, including subshells and pipelines.
	- Introduced tests for simple and complex subshell commands to validate their behavior.

- **Chores**
	- Updated file header comments to reflect the latest author and timestamp changes.

### CODE REVIEW

File: src/exec/handle_io.c [hunk 1]
```diff
@@ -86,7 +86,8 @@ static void	handle_pipeline(t_pipeline_conf *conf)
 
 void	handle_io(t_pipeline_conf *conf, t_list *redirects)
 {
-	handle_pipeline(conf);
+	if (conf)
```
Comment (alex216) [inline src/exec/handle_io.c:89]: conf is originally defined in `exec()` as a stack variable.
I am just curious, does that mean conf here always has a valid address, which will never be NULL? correct me if I am wrong!

Comment (rask24) [reply src/exec/handle_io.c:89]: The `if` statement is used for handling redirects in a subshell.
Subshells only support file redirection and do not support pipelines.
Therefore, in the `execute_subshell` function, the first argument of `handle_io` is set to NULL.

File: src/exec/wait_for_children.c [hunk 2]
```diff
@@ -15,22 +15,40 @@
...
-		{
-			if (WIFEXITED(status))
-				ctx->exit_status = WEXITSTATUS(status);
-			else if (WIFSIGNALED(status))
-				ctx->exit_status = WTERMSIG(status) + 128;
-		}
+		if (errno == ECHILD)
+			return (false);
+		print_error("waitpid", strerror(errno));
+		return (false);
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/exec/wait_for_children.c:37]: I've fixed it.

File: src/exec/wait_for_children.c [hunk 3]
```diff
@@ -15,22 +15,40 @@
...
 #include "ctx.h"
 
-void	wait_for_children(pid_t last_pid, t_ctx *ctx)
+static void	update_exit_status(int status, t_ctx *ctx)
+{
+	if (WIFEXITED(status))
+		ctx->exit_status = WEXITSTATUS(status);
+	else if (WIFSIGNALED(status))
+		ctx->exit_status = WTERMSIG(status) + 128;
+}
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/exec/wait_for_children.c:24]: This project doesn't implement jobs control. So it is not neccesary.

File: src/exec/execute_subshell.c [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	pid = fork();
+	if (pid == -1)
+	{
+		print_error("fork", strerror(errno));
+		return (EXIT_FAILURE);
+	}
+	if (pid == 0)
+	{
+		handle_io(conf, node->redirects, ctx, true);
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/exec/execute_subshell.c:33]: Exactly! I've fixed it.

File: tests/e2e/test_redirect.py [hunk 5]
```diff
@@ -244,6 +276,48 @@ def test_error_redirection_and_builtin(shell_session):
...
+            assert f.read() == "Hello\n"
+
+        assert os.path.exists(test_file2)
+        with open(test_file2, "r") as f:
+            assert f.read() == ""
+    finally:
+        for file in [test_file1, test_file2]:
+            if os.path.exists(file):
+                os.remove(file)
+
```
[the remark this answers is not in the document]
Comment (rask24) [reply tests/e2e/test_redirect.py:298]: I think this `finally` is correctly used. It's the pair of `try` not `with`.

File: src/exec/execute_pipeline.c [hunk 6]
```diff
@@ -22,53 +22,71 @@
...
-		return (EXIT_FAILURE);
+		if (pipe(pipe_fd) == -1)
+		{
+			print_error("pipe", strerror(errno));
+			return (free_and_return(conf, EXIT_FAILURE));
+		}
+		update_pipe_conf(conf, pipe_fd);
+		if (execute_ast_node(cur->left, ctx, conf) == EXIT_FAILURE)
+			return (free_and_return(conf, EXIT_FAILURE));
+		cur = cur->right;
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/exec/execute_pipeline.c:87]: `execute_ast_node(cur->left, ctx, conf)` will return EXIT_FAILURE only if `fork` fails.

We can consider the below 2 patterns.
1. execute_command, spawn_command
2. execute_subshell

So we have to close prev->read, next->write, next->read

**Review by alex216 — APPROVED** [review]

**Pull request merged**